# Enterprise Retail Intelligence Platform

## Notebook 03 — Data Cleaning

### Author
Shobha Saxena

### Objective

This notebook performs professional data cleaning on the extracted PostgreSQL datasets.

Cleaning tasks include:

- Missing value analysis
- Duplicate detection
- Data type conversion
- Text standardization
- Data validation
- Export cleaned datasets for EDA

In [1]:
# ============================================================
# Import Libraries
# ============================================================

import pandas as pd
import numpy as np

from pathlib import Path

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 20)

In [3]:
# ============================================================
# Load Extracted CSV Files
# ============================================================

from pathlib import Path

data_path = Path("../data/extracted")

customers_df = pd.read_csv(data_path / "customers.csv")
products_df = pd.read_csv(data_path / "products.csv")
orders_df = pd.read_csv(data_path / "orders.csv")
suppliers_df = pd.read_csv(data_path / "suppliers.csv")
inventory_df = pd.read_csv(data_path / "inventory.csv")
transportation_df = pd.read_csv(data_path / "transportation_logistics.csv")
payments_df = pd.read_csv(data_path / "payments.csv")
returns_df = pd.read_csv(data_path / "returns.csv")
customer_discounts_df = pd.read_csv(data_path / "customer_discounts.csv")

print("✅ All extracted datasets loaded successfully!")

✅ All extracted datasets loaded successfully!


In [4]:
# ============================================================
# Dataset Summary
# ============================================================

datasets = {
    "Customers": customers_df,
    "Products": products_df,
    "Orders": orders_df,
    "Suppliers": suppliers_df,
    "Inventory": inventory_df,
    "Transportation": transportation_df,
    "Payments": payments_df,
    "Returns": returns_df,
    "Customer Discounts": customer_discounts_df
}

summary = []

for name, df in datasets.items():
    summary.append({
        "Dataset": name,
        "Rows": df.shape[0],
        "Columns": df.shape[1]
    })

summary_df = pd.DataFrame(summary)

summary_df

,Dataset,Rows,Columns
0,Customers,793,4
1,Products,1861,5
2,Orders,9800,14
3,Suppliers,17,9
4,Inventory,1861,10
5,Transportation,9800,10
6,Payments,9800,8
7,Returns,490,8
8,Customer Discounts,793,8


In [5]:
# ============================================================
# Missing Value Analysis
# ============================================================

missing_summary = []

for name, df in datasets.items():
    missing = df.isnull().sum()
    missing = missing[missing > 0]

    if len(missing) > 0:
        for column, value in missing.items():
            missing_summary.append({
                "Dataset": name,
                "Column": column,
                "Missing Values": value,
                "Percentage": round((value / len(df)) * 100, 2)
            })

missing_df = pd.DataFrame(missing_summary)

missing_df

,Dataset,Column,Missing Values,Percentage
0,Orders,postal_code,11,0.11
1,Transportation,actual_delivery_date,953,9.72


## Missing Value Findings

### Orders

- `postal_code` contains **11 missing values (0.11%)**
- These values were intentionally left as NULL because verified postal codes were unavailable.

### Transportation

- `actual_delivery_date` contains **953 missing values (9.72%)**
- This is expected because some shipments have not yet been delivered.
- Missing values are retained as NULL.

In [6]:
# ============================================================
# Check Data Types
# ============================================================

orders_df.dtypes

row_id           int64
order_id        object
order_date      object
ship_date       object
ship_mode       object
customer_id     object
product_id      object
country         object
city            object
state           object
postal_code    float64
region          object
sales          float64
created_at      object
dtype: object

In [8]:
print(transportation_df.columns.tolist())

['transport_id', 'order_row_id', 'carrier_name', 'tracking_number', 'shipment_status', 'dispatch_date', 'estimated_delivery_date', 'actual_delivery_date', 'delivery_cost', 'created_at']


In [9]:
# ============================================================
# Convert Date Columns
# ============================================================

# Orders Table
orders_df["order_date"] = pd.to_datetime(orders_df["order_date"])
orders_df["ship_date"] = pd.to_datetime(orders_df["ship_date"])
orders_df["created_at"] = pd.to_datetime(orders_df["created_at"])

# Transportation Table
transportation_df["dispatch_date"] = pd.to_datetime(
    transportation_df["dispatch_date"]
)

transportation_df["estimated_delivery_date"] = pd.to_datetime(
    transportation_df["estimated_delivery_date"]
)

transportation_df["actual_delivery_date"] = pd.to_datetime(
    transportation_df["actual_delivery_date"]
)

transportation_df["created_at"] = pd.to_datetime(
    transportation_df["created_at"]
)

print("✅ Date columns converted successfully!")

✅ Date columns converted successfully!


In [10]:
orders_df.dtypes

row_id                  int64
order_id               object
order_date     datetime64[ns]
ship_date      datetime64[ns]
ship_mode              object
customer_id            object
product_id             object
country                object
city                   object
state                  object
postal_code           float64
region                 object
sales                 float64
created_at     datetime64[ns]
dtype: object

In [11]:
# ============================================================
# Duplicate Analysis
# ============================================================

duplicate_summary = []

for name, df in datasets.items():
    duplicate_summary.append({
        "Dataset": name,
        "Duplicate Rows": df.duplicated().sum()
    })

duplicate_df = pd.DataFrame(duplicate_summary)

duplicate_df

,Dataset,Duplicate Rows
0,Customers,0
1,Products,0
2,Orders,0
3,Suppliers,0
4,Inventory,0
5,Transportation,0
6,Payments,0
7,Returns,0
8,Customer Discounts,0


## Duplicate Analysis

### Objective
Check each extracted dataset for duplicate records before performing data preprocessing.

### Findings

- No duplicate records were detected in any dataset.
- Data integrity has been maintained during the ETL process.
- No duplicate removal was required.

### Conclusion

All datasets are suitable for further preprocessing and exploratory data analysis.

In [12]:
# ============================================================
# Data Type Conversion
# ============================================================

# ------------------------
# Orders Table
# ------------------------
orders_df["order_date"] = pd.to_datetime(orders_df["order_date"])
orders_df["ship_date"] = pd.to_datetime(orders_df["ship_date"])
orders_df["created_at"] = pd.to_datetime(orders_df["created_at"])

# ------------------------
# Transportation Table
# ------------------------
transportation_df["dispatch_date"] = pd.to_datetime(
    transportation_df["dispatch_date"]
)

transportation_df["estimated_delivery_date"] = pd.to_datetime(
    transportation_df["estimated_delivery_date"]
)

transportation_df["actual_delivery_date"] = pd.to_datetime(
    transportation_df["actual_delivery_date"]
)

transportation_df["created_at"] = pd.to_datetime(
    transportation_df["created_at"]
)

print("✅ Date columns converted successfully!")

✅ Date columns converted successfully!


In [13]:
# ============================================================
# Verify Data Types
# ============================================================

print("Orders Table")
print(orders_df.dtypes)

print("\n" + "="*50 + "\n")

print("Transportation Table")
print(transportation_df.dtypes)

Orders Table
row_id                  int64
order_id               object
order_date     datetime64[ns]
ship_date      datetime64[ns]
ship_mode              object
customer_id            object
product_id             object
country                object
city                   object
state                  object
postal_code           float64
region                 object
sales                 float64
created_at     datetime64[ns]
dtype: object


Transportation Table
transport_id                        int64
order_row_id                        int64
carrier_name                       object
tracking_number                    object
shipment_status                    object
dispatch_date              datetime64[ns]
estimated_delivery_date    datetime64[ns]
actual_delivery_date       datetime64[ns]
delivery_cost                     float64
created_at                 datetime64[ns]
dtype: object


## Data Type Conversion

### Objective

Convert all date-related columns from text (`object`) to proper `datetime` format.

### Converted Columns

**Orders**
- order_date
- ship_date
- created_at

**Transportation**
- dispatch_date
- estimated_delivery_date
- actual_delivery_date
- created_at

### Result

All date columns were successfully converted to `datetime64[ns]`, making them suitable for time-series analysis, feature engineering, and Power BI visualizations.

In [14]:
# ============================================================
# Text Standardization
# ============================================================

# ---------- Orders ----------
text_columns_orders = [
    "country",
    "city",
    "state",
    "region",
    "ship_mode"
]

for col in text_columns_orders:
    orders_df[col] = (
        orders_df[col]
        .astype(str)
        .str.strip()
        .str.title()
    )

# ---------- Transportation ----------
text_columns_transportation = [
    "carrier_name",
    "shipment_status"
]

for col in text_columns_transportation:
    transportation_df[col] = (
        transportation_df[col]
        .astype(str)
        .str.strip()
        .str.title()
    )

print("✅ Text standardization completed successfully!")

✅ Text standardization completed successfully!


In [15]:
orders_df[text_columns_orders].head()

,country,city,state,region,ship_mode
0,United States,Henderson,Kentucky,South,Second Class
1,United States,Henderson,Kentucky,South,Second Class
2,United States,Los Angeles,California,West,Second Class
3,United States,Fort Lauderdale,Florida,South,Standard Class
4,United States,Fort Lauderdale,Florida,South,Standard Class


In [16]:
transportation_df[text_columns_transportation].head()

,carrier_name,shipment_status
0,Delhivery,Delivered
1,Xpressbees,Delivered
2,Fedex,Delivered
3,Dhl,Delivered
4,Dhl,Delivered


In [17]:
# ============================================================
# Data Validation Report
# ============================================================

validation_report = []

for name, df in datasets.items():

    validation_report.append({
        "Dataset": name,
        "Rows": len(df),
        "Columns": len(df.columns),
        "Missing Values": df.isnull().sum().sum(),
        "Duplicate Rows": df.duplicated().sum()
    })

validation_df = pd.DataFrame(validation_report)

validation_df

,Dataset,Rows,Columns,Missing Values,Duplicate Rows
0,Customers,793,4,0,0
1,Products,1861,5,0,0
2,Orders,9800,14,11,0
3,Suppliers,17,9,0,0
4,Inventory,1861,10,0,0
5,Transportation,9800,10,953,0
6,Payments,9800,8,0,0
7,Returns,490,8,0,0
8,Customer Discounts,793,8,0,0


## Data Validation

### Validation Summary

The cleaned datasets were validated after preprocessing.

Validation checks included:

- Record count
- Column count
- Missing values
- Duplicate records

### Results

- No duplicate records detected.
- Expected missing values retained:
  - Orders → postal_code
  - Transportation → actual_delivery_date
- All date columns successfully converted.
- Text fields standardized.
- Datasets are ready for exploratory data analysis (EDA).

In [18]:
# ============================================================
# Save Cleaned Datasets
# ============================================================

from pathlib import Path

processed_path = Path("../data/processed")
processed_path.mkdir(parents=True, exist_ok=True)

customers_df.to_csv(processed_path / "customers.csv", index=False)
products_df.to_csv(processed_path / "products.csv", index=False)
orders_df.to_csv(processed_path / "orders.csv", index=False)
suppliers_df.to_csv(processed_path / "suppliers.csv", index=False)
inventory_df.to_csv(processed_path / "inventory.csv", index=False)
transportation_df.to_csv(processed_path / "transportation.csv", index=False)
payments_df.to_csv(processed_path / "payments.csv", index=False)
returns_df.to_csv(processed_path / "returns.csv", index=False)
customer_discounts_df.to_csv(processed_path / "customer_discounts.csv", index=False)

print("===================================================")
print("✅ All cleaned datasets saved successfully!")
print("===================================================")
print(processed_path.resolve())

✅ All cleaned datasets saved successfully!
D:\My Project\enterprise-retail-intelligence-platform-main\python\data\processed


In [19]:
# ============================================================
# Export Cleaned Data
# Enterprise Retail Intelligence Platform
# ============================================================

from pathlib import Path

# Create output directory if it does not exist
processed_path = Path("../data/processed")
processed_path.mkdir(parents=True, exist_ok=True)

# Dictionary containing cleaned DataFrames
processed_datasets = {
    "customers": customers_df,
    "products": products_df,
    "orders": orders_df,
    "suppliers": suppliers_df,
    "inventory": inventory_df,
    "transportation": transportation_df,
    "payments": payments_df,
    "returns": returns_df,
    "customer_discounts": customer_discounts_df,
}

# Export each dataset
for name, dataframe in processed_datasets.items():
    dataframe.to_csv(processed_path / f"{name}.csv", index=False)

print("=" * 60)
print("SUCCESS: All cleaned datasets exported successfully.")
print("=" * 60)
print(f"Output Location : {processed_path.resolve()}")
print(f"Datasets Exported : {len(processed_datasets)}")

SUCCESS: All cleaned datasets exported successfully.
Output Location : D:\My Project\enterprise-retail-intelligence-platform-main\python\data\processed
Datasets Exported : 9


In [20]:
# ============================================================
# Verify Exported Files
# ============================================================

exported_files = sorted(processed_path.glob("*.csv"))

print("Processed Files\n")

for file in exported_files:
    print(f"✓ {file.name}")

print("\nTotal Files :", len(exported_files))

Processed Files

✓ customer_discounts.csv
✓ customers.csv
✓ inventory.csv
✓ orders.csv
✓ payments.csv
✓ products.csv
✓ returns.csv
✓ superstore_cleaned.csv
✓ suppliers.csv
✓ transportation.csv

Total Files : 10


# Data Cleaning Summary

## Objective

Prepare all extracted database tables for exploratory data analysis, feature engineering, SQL reporting, and Power BI dashboard development.

---

## Cleaning Steps Performed

- Imported extracted datasets
- Verified dataset dimensions
- Checked missing values
- Verified duplicate records
- Converted date columns to datetime
- Standardized text fields
- Validated cleaned datasets
- Exported processed datasets

---

## Output Directory

python/data/processed/

---

## Next Notebook

04_exploratory_data_analysis.ipynb

The processed datasets generated in this notebook will be used for exploratory data analysis, business insights generation, and dashboard development.